{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import logging\n",
    "import os\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "from napistu import utils\n",
    "from napistu.gcs import downloads\n",
    "\n",
    "# Set up logging\n",
    "logging.basicConfig(level=logging.DEBUG)\n",
    "logger = logging.getLogger(\"napistu\")\n",
    "\n",
    "import httpx\n",
    "\n",
    "NAPISTU_DATA_DIR = os.path.expanduser(\"~/Desktop/DATA/napistu_data\")\n",
    "NAPISTU_ASSET = \"human_consensus_w_distances\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 0/4 [00:00<?, ?it/s]DEBUG:napistu.consensus:Building consensus identifiers for species\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "DEBUG:napistu.consensus:Creating lookup table for species\n",
      "DEBUG:napistu.utils:label is not defined in table_schema; adding a constant (1)\n",
      "DEBUG:napistu.consensus:Preparing consensus table for species\n",
      "DEBUG:napistu.consensus:Validating consensus table for species\n",
      "DEBUG:napistu.consensus:Building consensus identifiers for species\n",
      "DEBUG:napistu.consensus:Creating lookup table for species\n",
      "DEBUG:napistu.utils:label is not defined in table_schema; adding a constant (1)\n",
      "DEBUG:napistu.consensus:Preparing consensus table for species\n",
      "DEBUG:napistu.consensus:Validating consensus table for species\n",
      "DEBUG:napistu.consensus:Building consensus identifiers for species\n",
      "DEBUG:napistu.consensus:Creating lookup table for species\n",
      "DEBUG:napistu.utils:label is not defined in table_schema; adding a constant (1)\n",
      "DEBUG:napistu.consensus:Preparing consensus table for species\n",
      "DEBUG:napistu.consensus:Validating consensus table for species\n",
      " 75%|███████▌  | 3/4 [00:00<00:00, 19.92it/s]DEBUG:napistu.consensus:Building consensus identifiers for species\n",
      "DEBUG:napistu.consensus:Creating lookup table for species\n",
      "DEBUG:napistu.utils:label is not defined in table_schema; adding a constant (1)\n",
      "DEBUG:napistu.consensus:Preparing consensus table for species\n",
      "DEBUG:napistu.consensus:Validating consensus table for species\n",
      "100%|██████████| 4/4 [00:00<00:00, 16.44it/s]\n",
      "INFO:napistu.consensus:Reporting possible issues in component models\n",
      "INFO:napistu.consensus:Evaluating compartment compatibility\n",
      "DEBUG:napistu.consensus:Building consensus identifiers for compartments\n",
      "DEBUG:napistu.consensus:Creating lookup table for compartments\n",
      "DEBUG:napistu.consensus:Preparing consensus table for compartments\n",
      "INFO:napistu.consensus:Creating source table\n",
      "INFO:napistu.consensus:Aggregating old sources\n",
      "INFO:napistu.consensus:Returning new source table\n",
      "DEBUG:napistu.consensus:Validating consensus table for compartments\n",
      "INFO:napistu.consensus:>>>> 14 c_id entries merged into 6\n",
      "INFO:napistu.consensus:Testing for identical formulas of to-be-merged reactions\n",
      "INFO:napistu.consensus:All merges names matched exactly\n",
      "INFO:napistu.consensus:Evaluating ontology compatibility\n",
      "DEBUG:napistu.sbml_dfs_core:loading all identifiers\n",
      "DEBUG:napistu.sbml_dfs_core:loading all identifiers\n",
      "DEBUG:napistu.sbml_dfs_core:loading all identifiers\n",
      "DEBUG:napistu.sbml_dfs_core:loading all identifiers\n",
      "DEBUG:napistu.consensus:Pre-consensus checks complete\n",
      "INFO:napistu.sbml_dfs_utils:Running in dogmatic mode - differences genes, transcripts, and proteins will try to be maintained as separate species.\n",
      "INFO:napistu.consensus:Defining compartments based on unique ids\n",
      "DEBUG:napistu.consensus:Building consensus identifiers for compartments\n",
      "DEBUG:napistu.consensus:Creating lookup table for compartments\n",
      "DEBUG:napistu.consensus:Preparing consensus table for compartments\n",
      "INFO:napistu.consensus:Creating source table\n",
      "INFO:napistu.consensus:Aggregating old sources\n",
      "INFO:napistu.consensus:Returning new source table\n",
      "DEBUG:napistu.consensus:Validating consensus table for compartments\n",
      "INFO:napistu.consensus:>>>> 14 c_id entries merged into 6\n",
      "INFO:napistu.consensus:Testing for identical formulas of to-be-merged reactions\n",
      "INFO:napistu.consensus:All merges names matched exactly\n",
      "INFO:napistu.consensus:Defining species based on unique ids\n",
      "DEBUG:napistu.consensus:Building consensus identifiers for species\n",
      "DEBUG:napistu.consensus:Creating lookup table for species\n",
      "DEBUG:napistu.consensus:Preparing consensus table for species\n",
      "INFO:napistu.consensus:Creating source table\n",
      "INFO:napistu.consensus:Aggregating old sources\n",
      "INFO:napistu.consensus:Returning new source table\n",
      "DEBUG:napistu.consensus:Validating consensus table for species\n",
      "INFO:napistu.consensus:>>>> 148 s_id entries merged into 64\n",
      "INFO:napistu.consensus:Testing for identical formulas of to-be-merged reactions\n",
      "INFO:napistu.consensus:All merges names matched exactly\n",
      "INFO:napistu.consensus:Defining compartmentalized species based on unique species x compartments\n",
      "INFO:napistu.consensus:>>>> 164 sc_id entries merged into 77\n",
      "INFO:napistu.consensus:Testing for identical formulas of to-be-merged reactions\n",
      "WARNING:napistu.consensus:\n",
      "1 merges were of entities with distinct names, including:\n",
      "\n",
      "WARNING:napistu.consensus:PRPS2 dimer, PRPS1L dimer [cytosol] & PRPS1 dimer [cytosol]\n",
      "\n",
      "INFO:napistu.consensus:Creating source table\n",
      "INFO:napistu.consensus:Aggregating old sources\n",
      "INFO:napistu.consensus:Returning new source table\n",
      "INFO:napistu.consensus:Define reactions based on membership of identical compartmentalized species\n",
      "INFO:napistu.consensus:Merging reactions based on identical membership (sc_id + stoichiometry)\n",
      "INFO:napistu.consensus:>>>> 50 r_id entries merged into 25\n",
      "INFO:napistu.consensus:Creating formulas for to-be-merged reactions to help with reporting merges of reactions with inconsistently named reactants\n",
      "INFO:napistu.consensus:Done creating reaction formulas\n",
      "INFO:napistu.consensus:Testing for identical formulas of to-be-merged reactions\n",
      "INFO:napistu.consensus:All merges names matched exactly\n",
      "INFO:napistu.consensus:Merging reactions identifiers\n",
      "INFO:napistu.consensus:Merging reactions sources\n",
      "INFO:napistu.consensus:Creating source table\n",
      "INFO:napistu.consensus:Aggregating old sources\n",
      "INFO:napistu.consensus:Returning new source table\n",
      "INFO:napistu.consensus:Annotating reversibility based on merged reactions\n",
      "INFO:napistu.consensus:Define reaction species based on reactions\n",
      "DEBUG:napistu.utils:label is not defined in table_schema; adding a constant (1)\n",
      "WARNING:napistu.sbml_dfs_utils:<2 priority pathways found, using all pathways. Set priority_pathways as None explicitly to remove this warning.\n",
      "WARNING:napistu.sbml_dfs_utils:<2 priority pathways found, using all pathways. Set priority_pathways as None explicitly to remove this warning.\n",
      "DEBUG:napistu.sbml_dfs_core:loading all identifiers\n",
      "WARNING:napistu.sbml_dfs_utils:<2 priority pathways found, using all pathways. Set priority_pathways as None explicitly to remove this warning.\n",
      "WARNING:napistu.sbml_dfs_utils:<2 priority pathways found, using all pathways. Set priority_pathways as None explicitly to remove this warning.\n",
      "DEBUG:napistu.sbml_dfs_core:loading all identifiers\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'Entity type: species, Check type: source_cooccurrence'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "#T_af1b7 th {\n",
       "  color: limegreen;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_af1b7\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"index_name level0\" >pathway_id</th>\n",
       "      <th id=\"T_af1b7_level0_col0\" class=\"col_heading level0 col0\" >glucose metabolism</th>\n",
       "      <th id=\"T_af1b7_level0_col1\" class=\"col_heading level0 col1\" >glycolysis</th>\n",
       "      <th id=\"T_af1b7_level0_col2\" class=\"col_heading level0 col2\" >ppp</th>\n",
       "      <th id=\"T_af1b7_level0_col3\" class=\"col_heading level0 col3\" >tca</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th class=\"index_name level0\" >pathway_id</th>\n",
       "      <th class=\"blank col0\" >&nbsp;</th>\n",
       "      <th class=\"blank col1\" >&nbsp;</th>\n",
       "      <th class=\"blank col2\" >&nbsp;</th>\n",
       "      <th class=\"blank col3\" >&nbsp;</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th id=\"T_af1b7_level0_row0\" class=\"row_heading level0 row0\" >glucose metabolism</th>\n",
       "      <td id=\"T_af1b7_row0_col0\" class=\"data row0 col0\" >80</td>\n",
       "      <td id=\"T_af1b7_row0_col1\" class=\"data row0 col1\" >55</td>\n",
       "      <td id=\"T_af1b7_row0_col2\" class=\"data row0 col2\" >10</td>\n",
       "      <td id=\"T_af1b7_row0_col3\" class=\"data row0 col3\" >17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_af1b7_level0_row1\" class=\"row_heading level0 row1\" >glycolysis</th>\n",
       "      <td id=\"T_af1b7_row1_col0\" class=\"data row1 col0\" >55</td>\n",
       "      <td id=\"T_af1b7_row1_col1\" class=\"data row1 col1\" >55</td>\n",
       "      <td id=\"T_af1b7_row1_col2\" class=\"data row1 col2\" >9</td>\n",
       "      <td id=\"T_af1b7_row1_col3\" class=\"data row1 col3\" >10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_af1b7_level0_row2\" class=\"row_heading level0 row2\" >ppp</th>\n",
       "      <td id=\"T_af1b7_row2_col0\" class=\"data row2 col0\" >10</td>\n",
       "      <td id=\"T_af1b7_row2_col1\" class=\"data row2 col1\" >9</td>\n",
       "      <td id=\"T_af1b7_row2_col2\" class=\"data row2 col2\" >41</td>\n",
       "      <td id=\"T_af1b7_row2_col3\" class=\"data row2 col3\" >8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_af1b7_level0_row3\" class=\"row_heading level0 row3\" >tca</th>\n",
       "      <td id=\"T_af1b7_row3_col0\" class=\"data row3 col0\" >17</td>\n",
       "      <td id=\"T_af1b7_row3_col1\" class=\"data row3 col1\" >10</td>\n",
       "      <td id=\"T_af1b7_row3_col2\" class=\"data row3 col2\" >8</td>\n",
       "      <td id=\"T_af1b7_row3_col3\" class=\"data row3 col3\" >37</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x143899e10>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'Entity type: species, Check type: ontology_x_source_cooccurrence'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "#T_a63fd th {\n",
       "  color: limegreen;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_a63fd\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"index_name level0\" >pathway_id</th>\n",
       "      <th id=\"T_a63fd_level0_col0\" class=\"col_heading level0 col0\" >glucose metabolism</th>\n",
       "      <th id=\"T_a63fd_level0_col1\" class=\"col_heading level0 col1\" >glycolysis</th>\n",
       "      <th id=\"T_a63fd_level0_col2\" class=\"col_heading level0 col2\" >ppp</th>\n",
       "      <th id=\"T_a63fd_level0_col3\" class=\"col_heading level0 col3\" >tca</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th class=\"index_name level0\" >bqb_ontology</th>\n",
       "      <th class=\"blank col0\" >&nbsp;</th>\n",
       "      <th class=\"blank col1\" >&nbsp;</th>\n",
       "      <th class=\"blank col2\" >&nbsp;</th>\n",
       "      <th class=\"blank col3\" >&nbsp;</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th id=\"T_a63fd_level0_row0\" class=\"row_heading level0 row0\" >BQB_HAS_PART::chebi</th>\n",
       "      <td id=\"T_a63fd_row0_col0\" class=\"data row0 col0\" >8</td>\n",
       "      <td id=\"T_a63fd_row0_col1\" class=\"data row0 col1\" >6</td>\n",
       "      <td id=\"T_a63fd_row0_col2\" class=\"data row0 col2\" >4</td>\n",
       "      <td id=\"T_a63fd_row0_col3\" class=\"data row0 col3\" >7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_a63fd_level0_row1\" class=\"row_heading level0 row1\" >BQB_HAS_PART::uniprot</th>\n",
       "      <td id=\"T_a63fd_row1_col0\" class=\"data row1 col0\" >32</td>\n",
       "      <td id=\"T_a63fd_row1_col1\" class=\"data row1 col1\" >23</td>\n",
       "      <td id=\"T_a63fd_row1_col2\" class=\"data row1 col2\" >7</td>\n",
       "      <td id=\"T_a63fd_row1_col3\" class=\"data row1 col3\" >13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_a63fd_level0_row2\" class=\"row_heading level0 row2\" >BQB_IS::chebi</th>\n",
       "      <td id=\"T_a63fd_row2_col0\" class=\"data row2 col0\" >38</td>\n",
       "      <td id=\"T_a63fd_row2_col1\" class=\"data row2 col1\" >30</td>\n",
       "      <td id=\"T_a63fd_row2_col2\" class=\"data row2 col2\" >29</td>\n",
       "      <td id=\"T_a63fd_row2_col3\" class=\"data row2 col3\" >23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_a63fd_level0_row3\" class=\"row_heading level0 row3\" >BQB_IS::reactome</th>\n",
       "      <td id=\"T_a63fd_row3_col0\" class=\"data row3 col0\" >80</td>\n",
       "      <td id=\"T_a63fd_row3_col1\" class=\"data row3 col1\" >55</td>\n",
       "      <td id=\"T_a63fd_row3_col2\" class=\"data row3 col2\" >41</td>\n",
       "      <td id=\"T_a63fd_row3_col3\" class=\"data row3 col3\" >37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_a63fd_level0_row4\" class=\"row_heading level0 row4\" >BQB_IS::uniprot</th>\n",
       "      <td id=\"T_a63fd_row4_col0\" class=\"data row4 col0\" >10</td>\n",
       "      <td id=\"T_a63fd_row4_col1\" class=\"data row4 col1\" >2</td>\n",
       "      <td id=\"T_a63fd_row4_col2\" class=\"data row4 col2\" >5</td>\n",
       "      <td id=\"T_a63fd_row4_col3\" class=\"data row4 col3\" >1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_a63fd_level0_row5\" class=\"row_heading level0 row5\" >BQB_IS_DESCRIBED_BY::pubmed</th>\n",
       "      <td id=\"T_a63fd_row5_col0\" class=\"data row5 col0\" >9</td>\n",
       "      <td id=\"T_a63fd_row5_col1\" class=\"data row5 col1\" >7</td>\n",
       "      <td id=\"T_a63fd_row5_col2\" class=\"data row5 col2\" >4</td>\n",
       "      <td id=\"T_a63fd_row5_col3\" class=\"data row5 col3\" >2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_a63fd_level0_row6\" class=\"row_heading level0 row6\" >BQB_IS_HOMOLOG_TO::reactome</th>\n",
       "      <td id=\"T_a63fd_row6_col0\" class=\"data row6 col0\" >41</td>\n",
       "      <td id=\"T_a63fd_row6_col1\" class=\"data row6 col1\" >24</td>\n",
       "      <td id=\"T_a63fd_row6_col2\" class=\"data row6 col2\" >12</td>\n",
       "      <td id=\"T_a63fd_row6_col3\" class=\"data row6 col3\" >14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x14387a990>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'Entity type: compartments, Check type: source_cooccurrence'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "#T_04984 th {\n",
       "  color: limegreen;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_04984\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"index_name level0\" >pathway_id</th>\n",
       "      <th id=\"T_04984_level0_col0\" class=\"col_heading level0 col0\" >glucose metabolism</th>\n",
       "      <th id=\"T_04984_level0_col1\" class=\"col_heading level0 col1\" >glycolysis</th>\n",
       "      <th id=\"T_04984_level0_col2\" class=\"col_heading level0 col2\" >ppp</th>\n",
       "      <th id=\"T_04984_level0_col3\" class=\"col_heading level0 col3\" >tca</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th class=\"index_name level0\" >pathway_id</th>\n",
       "      <th class=\"blank col0\" >&nbsp;</th>\n",
       "      <th class=\"blank col1\" >&nbsp;</th>\n",
       "      <th class=\"blank col2\" >&nbsp;</th>\n",
       "      <th class=\"blank col3\" >&nbsp;</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th id=\"T_04984_level0_row0\" class=\"row_heading level0 row0\" >glucose metabolism</th>\n",
       "      <td id=\"T_04984_row0_col0\" class=\"data row0 col0\" >8</td>\n",
       "      <td id=\"T_04984_row0_col1\" class=\"data row0 col1\" >4</td>\n",
       "      <td id=\"T_04984_row0_col2\" class=\"data row0 col2\" >1</td>\n",
       "      <td id=\"T_04984_row0_col3\" class=\"data row0 col3\" >3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_04984_level0_row1\" class=\"row_heading level0 row1\" >glycolysis</th>\n",
       "      <td id=\"T_04984_row1_col0\" class=\"data row1 col0\" >4</td>\n",
       "      <td id=\"T_04984_row1_col1\" class=\"data row1 col1\" >4</td>\n",
       "      <td id=\"T_04984_row1_col2\" class=\"data row1 col2\" >1</td>\n",
       "      <td id=\"T_04984_row1_col3\" class=\"data row1 col3\" >1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_04984_level0_row2\" class=\"row_heading level0 row2\" >ppp</th>\n",
       "      <td id=\"T_04984_row2_col0\" class=\"data row2 col0\" >1</td>\n",
       "      <td id=\"T_04984_row2_col1\" class=\"data row2 col1\" >1</td>\n",
       "      <td id=\"T_04984_row2_col2\" class=\"data row2 col2\" >1</td>\n",
       "      <td id=\"T_04984_row2_col3\" class=\"data row2 col3\" >1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_04984_level0_row3\" class=\"row_heading level0 row3\" >tca</th>\n",
       "      <td id=\"T_04984_row3_col0\" class=\"data row3 col0\" >3</td>\n",
       "      <td id=\"T_04984_row3_col1\" class=\"data row3 col1\" >1</td>\n",
       "      <td id=\"T_04984_row3_col2\" class=\"data row3 col2\" >1</td>\n",
       "      <td id=\"T_04984_row3_col3\" class=\"data row3 col3\" >4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x14387bd10>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'Entity type: compartments, Check type: ontology_x_source_cooccurrence'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "#T_2e94e th {\n",
       "  color: limegreen;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_2e94e\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"index_name level0\" >pathway_id</th>\n",
       "      <th id=\"T_2e94e_level0_col0\" class=\"col_heading level0 col0\" >glucose metabolism</th>\n",
       "      <th id=\"T_2e94e_level0_col1\" class=\"col_heading level0 col1\" >glycolysis</th>\n",
       "      <th id=\"T_2e94e_level0_col2\" class=\"col_heading level0 col2\" >ppp</th>\n",
       "      <th id=\"T_2e94e_level0_col3\" class=\"col_heading level0 col3\" >tca</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th class=\"index_name level0\" >bqb_ontology</th>\n",
       "      <th class=\"blank col0\" >&nbsp;</th>\n",
       "      <th class=\"blank col1\" >&nbsp;</th>\n",
       "      <th class=\"blank col2\" >&nbsp;</th>\n",
       "      <th class=\"blank col3\" >&nbsp;</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th id=\"T_2e94e_level0_row0\" class=\"row_heading level0 row0\" >BQB_IS::go</th>\n",
       "      <td id=\"T_2e94e_row0_col0\" class=\"data row0 col0\" >8</td>\n",
       "      <td id=\"T_2e94e_row0_col1\" class=\"data row0 col1\" >4</td>\n",
       "      <td id=\"T_2e94e_row0_col2\" class=\"data row0 col2\" >1</td>\n",
       "      <td id=\"T_2e94e_row0_col3\" class=\"data row0 col3\" >4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x140688990>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "DEBUG:napistu.network.net_create:DEBUG: creating compartmentalized species features\n",
      "INFO:napistu.network.net_create:Organizing all network nodes (compartmentalized species and reactions)\n",
      "INFO:napistu.network.net_create:Formatting edges as a regulatory graph\n",
      "INFO:napistu.network.net_create_utils:Processing 434 reaction species using the regulatory hierarchy\n",
      "INFO:napistu.network.net_create:Adding reversibility and other meta-data from reactions_data\n",
      "INFO:napistu.network.net_create:Creating reverse reactions for reversible reactions on a directed graph\n",
      "WARNING:napistu.network.net_create:3 edges were dropped due to duplicated origin -> target relationiships, use verbose for more information\n",
      "INFO:napistu.network.net_create:Formatting NapistuGraph output\n",
      "INFO:napistu.network.ng_core:No isolated vertices found to remove\n"
     ]
    }
   ],
   "source": [
    "from napistu import indices, consensus, sbml_dfs_utils\n",
    "from napistu.network import net_create\n",
    "\n",
    "TESTDATA_PATH = \"../lib/napistu-py/src/tests/test_data\"\n",
    "\n",
    "pw_index = indices.PWIndex(os.path.join(TESTDATA_PATH, \"pw_index_metabolism.tsv\"))\n",
    "sbml_dfs_dict = consensus.construct_sbml_dfs_dict(pw_index)\n",
    "\n",
    "sbml_dfs = consensus.construct_consensus_model(sbml_dfs_dict, pw_index)\n",
    "sbml_dfs_utils.display_post_consensus_checks(sbml_dfs.post_consensus_checks())\n",
    "napistu_graph = net_create.create_napistu_graph(sbml_dfs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "entity_type = \"reactions\"\n",
    "\n",
    "def entity_data_table_to_dict(sbml_dfs, entity_type: str, table_name: str) -> dict:    \n",
    "    \n",
    "    entity_data = sbml_dfs._get_entity_data(entity_type, table_name)\n",
    "    \n",
    "    attrs = {}\n",
    "    for c in entity_data.columns:\n",
    "        attr_name = table_name + \"_\" + c\n",
    "        \n",
    "        attrs[attr_name] = {\n",
    "            WEIGHTING_SPEC.TABLE: table_name,\n",
    "            WEIGHTING_SPEC.VARIABLE: c,\n",
    "            WEIGHTING_SPEC.TRANSFORMATION: \"identity\"\n",
    "        }\n",
    "    \n",
    "    return attrs\n",
    "\n",
    "def entity_data_to_dict(sbml_dfs, entity_type: str, table_names: Optional[list[str]] = None) -> dict:\n",
    "\n",
    "    entity_data_attr = ENTITIES_TO_ENTITY_DATA[entity_type]\n",
    "    entity_data = getattr(sbml_dfs, entity_data_attr)\n",
    "\n",
    "    if len(entity_data) == 0:\n",
    "        raise ValueError(f\"No tables in {entity_data_attr}\")\n",
    "\n",
    "    if table_names is None:\n",
    "        table_names = entity_data.keys()\n",
    "    else:\n",
    "        invalid_tables = set(table_names) - set(entity_data.keys())\n",
    "        if invalid_tables:\n",
    "            raise ValueError(f\"The following requested tables do not exist in {entity_data_attr}: {invalid_tables}\")\n",
    "\n",
    "    entity_data_dicts = {}\n",
    "    for table_name in table_names:\n",
    "        entity_data_dicts[table_name] = entity_data_table_to_dict(sbml_dfs, entity_type, table_name)\n",
    "\n",
    "    # flatten dict\n",
    "    cross_table_entity_dict = {k: v for d in entity_data_dicts.values() for k, v in d.items()}\n",
    "\n",
    "    return cross_table_entity_dict\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# augment edges\n",
    "napistu_graph.set_graph_attrs({\n",
    "    \"reactions\" : entity_data_to_dict(sbml_dfs, SBML_DFS.REACTIONS)\n",
    "}, mode = \"extend\")\n",
    "\n",
    "# add the new attributes\n",
    "napistu_graph.add_edge_data(sbml_dfs, mode = \"extend\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
